python/plugins/processing/algs/qgis/Climb.py

```diff
@@ -0,0 +1,246 @@
+            return
+
+        thefields = QgsFields()
+        climbindex = -1
+        descentindex = -1
+        fieldnumber = 0
+
+        # Skip fields with names that are equal to the generated ones
+        for field in source.fields():
+            if str(field.name()) == str(self.CLIMBATTRIBUTE):
+                feedback.pushInfo("Warning: existing " +
+                                  str(self.CLIMBATTRIBUTE) +
+                                  " attribute found and removed")
+                climbindex = fieldnumber
+            elif str(field.name()) == str(self.DESCENTATTRIBUTE):
+                feedback.pushInfo("Warning: existing " +
+                                  str(self.DESCENTATTRIBUTE) +
+                                  " attribute found and removed")
+                descentindex = fieldnumber
+            else:
+                thefields.append(field)
+            fieldnumber = fieldnumber + 1
+
+        # Create new fields for climb and descent
+        thefields.append(QgsField(self.CLIMBATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.DESCENTATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MINELEVATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MAXELEVATTRIBUTE, QVariant.Double))
+
+        layerwithz = source
+
+        (sink, dest_id) = self.parameterAsSink(parameters,
+                                               self.OUTPUT,
+                                               context, thefields,
+                                               layerwithz.wkbType(),
+                                               source.sourceCrs())
+
+        # get features from source (with z values)
+        features = layerwithz.getFeatures()
+        totalclimb = 0
+        totaldescent = 0
+        minelevation = float('Infinity')
+        maxelevation = float('-Infinity')
+
+        for current, feature in enumerate(features):
+            # Stop the algorithm if cancelled
+            if feedback.isCanceled():
+                break
+            climb = 0
+            descent = 0
+            minelev = float('Infinity')
+            maxelev = float('-Infinity')
+            # In case of multigeometries we need to do the parts
+            for part in feature.geometry().constParts():
+                # Calculate the climb
+                first = True
+                zval = 0
+                for v in part.vertices():
+                    zval = v.z()
+                    if math.isnan(zval):
+                        feedback.pushInfo("Missing Z value")
+                        continue
+                    if first:
+                        prevz = zval
+                        minelev = zval
+                        maxelev = zval
+                        first = False
+                    else:
+                        diff = zval - prevz
```

Comment:
I sould add a column also with % repect part length not only delta Z.

Reply:
I don't understand what you mean.  Could you explain it another way?

Reply:
@havatv delta Z is not a useful information if not related how steep is related with length of the geometry part... e.g. 100m in 2km lenght is a different steep respect 100m in 100m lenght => https://en.wikipedia.org/wiki/Grade_(slope) . I would suggest to process also % not only climb. I agree that can be post processed later, but IMHO is something useful. The problem is that climb is the integration of all delta Z of the geometry parts and does not represent a real % but a medium %.

Reply:
@luipir, the purpose of this algorithm is to calculate the total / accumulated climb (and descent) for a line. So it is the delta Zs that provide useful information (the positive deltas contribute to climb, the negative deltas contribute to descent). An example of use is to calculate the total climb of a cycling race, a cross country racing ski track or a private training session.
I think that climb / descent as a percentage seldom would be of interest. A line will in general have some descending parts and some ascending parts, and I can't see what an average would be useful for (it would also be very easy to calculate from the numbers provided by this algorithm).
A thing that could be of intereste is the steepness/slope of the steepest ascent / descent.